### PR TITLE
docs: add specification of zkvm prover

### PR DESCRIPTION
**Pull request by dongchangYoo**
In this PR, the specification of zkvm prover was added to `./specs/experimental/zk-fault-dispute-game/overview.md`

It would be great to add appropriate links to the terms 'PreImage' and 'Dispute Game Factory' later.

### CODE REVIEW

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 1]
```diff
@@ -1,15 +1,28 @@
...
 <!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->
 **Table of Contents**
 
-- [Overview](#overview)
-- [ZK Fault Dispute Game Creation](#zk-fault-dispute-game-creation)
-- [Challenge Creation](#challenge-creation)
-- [Dissection](#dissection)
-- [Proving Fault using ZKVM](#proving-fault-using-zkvm)
-- [Resolution of ZK Fault Dispute Game](#resolution-of-zk-fault-dispute-game)
+- [ZK Fault Dispute Game](#zk-fault-dispute-game)
```
Comment (dongchangYoo) [inline specs/experimental/zk-fault-dipute-game/overview.md]: I'm not sure why these changes occurred. The changes were made by the `lint-specs-md-fix`, but should i revert them? @0xHansLee

**Comment by 0xHansLee** [comment]
@seolaoh and @sm-stack Could you review this PR?
I think the last touch for overall spec of ZK Fault Dispute Game is required, so please review the description of each content. 
I also plan to add `proveFault` interface to `Proving Fault using ZKVM` section.

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 1 above)
Comment (0xHansLee) [reply specs/experimental/zk-fault-dipute-game/overview.md]: This may be due to the header I manually added, `# ZK Fault Dispute Game`, so please ignore this.
I'll make appropriate changes in the last touch.

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 2]
```diff
@@ -23,4 +36,71 @@
...
+
+The zkVM (Zero-Knowledge Virtual Machine) is a virtual machine that executes guest program compiled with a specified
+compiler generating zero-knowledge proofs for their execution. The guest program can be written in standard programming
+languages, such as Rust and Go.
+
+#### Guest Program
+
+In the case of the ZK fault proof, the guest program is an extension of [L2 Chain Derivation][g-l2-chain-derivation]
+that includes a connectivity check among the blocks from L1 origin block and the specified block `C`. The hash of
+Block `C` is determined as the parent hash stored when the challenger initialized the [Dispute Game][g-dispute-game]
```
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md]: To be clarify, the dispute game is created and initialized by asserter who is game creator of dispute game. And challenger will create challenge against the claim if he/she think the submitted claim of dispute game is invalid. 
```suggestion
Block `C` is determined as the parent hash stored when the [Dispute Game][g-dispute-game] is created in Dispute Game Factory
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 3]
```diff
@@ -23,4 +36,71 @@
...
+by calling `create()` of the Dispute Game Factory.
+If the guest program consists only of L2 Chain Derivation, the challenger could create a proof with transactions that
+are not the actual transactions of the target block by manipulating an L1 Block that contains an
+[g-sequencer-batch][g-sequencer-batch].
```
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md]: This seems awkward in context. Rather, why not add in L59 that the L2 Chain Derivation and connectivity checks guarantee that it executes the actual transaction posted by sequencer?

**Review by 0xHansLee — COMMENTED** [review]
Could you add an interface or spec for request of proof generation to prover?

**Review by 0xHansLee — COMMENTED** [review]
I think I can add something between `## Proving Fault using ZKVM` and `### Prover` for context, such as that challenger need to generate ZK proof and submit on-chain to prove that the claim of the dispute game is fault. And also, interface for requesting ZK proof generation.

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 4]
```diff
@@ -1,15 +1,29 @@
 # ZK Fault Dispute Game
+<!-- All glossary references in this file. -->
+
+[g-zk-fault-proof]: ../../glossary.md#zk-fault-proof
+[g-l2-chain-derivation]: ../../glossary.md#l2-chain-derivation
+[g-sequencer-batch]: ../../glossary.md#sequencer-batch
+[g-dispute-game]: ../../experimental/zk-fault-dipute-game/overview.md
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]: This link just represents this file, so I think it's better to add it later when related terms are added to `glossary.md`.
```suggestion
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 5]
```diff
@@ -23,4 +37,91 @@
 
 ## Proving Fault using ZKVM
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]: Isn't zkVM natural than ZKVM?
```suggestion
## Proving Fault using zkVM
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 6]
```diff
@@ -23,4 +37,91 @@
...
+      bytes calldata _proofBytes
+    ) public view returns (
+      bytes32 parentOutputRoot, 
+      bytes32 outputRoot, 
+      bytes32 L1EndHash
+    );
+}
+```
+
+## Prover as an RPC Server
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]: Maybe this title level is more proper for this section.
```suggestion
### Prover as an RPC Server
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 7]
```diff
@@ -23,4 +37,91 @@
...
+
+### zkVM Proof
+
+The zkVM (Zero-Knowledge Virtual Machine) is a virtual machine that executes guest program compiled with a specified
+compiler generating zero-knowledge proofs for their execution. The guest program can be written in standard programming
+languages, such as Rust and Go.
+
+#### Guest Program
+
+In the case of the ZK fault proof, the guest program is an extension of [L2 Chain Derivation][g-l2-chain-derivation]
```
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
For ZK Fault Dispute Game, the guest program is an extension of [L2 Chain Derivation][g-l2-chain-derivation]
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 8]
```diff
@@ -23,4 +37,91 @@
...
+that includes a connectivity check among the blocks from L1 origin block and the specified block `C`. The hash of
```
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
that includes a connectivity check among the blocks from L1 origin block to the specified block `C`. The hash of
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 9]
```diff
@@ -23,4 +37,91 @@
...
+manipulating attack be thwarted by checking the block hash `C` value. For example, if an attacker creates a proof
+using a transaction that is not included in the [sequencer-batch][g-sequencer-batch], the block `C` value will change,
+preventing them from winning the challenge.
+
+#### Host Program
+
+The host program is a main part of the prover, responsible for delivering the guest program and its input to the zkVM.
+The host program first executes the guest program logic to gather the necessary data, which is equal to what is stored
+in the [PreImageOracle] during the [Optimism's Dispute Game], but the block headers from the L1 block containing
+the [sequencer-batch][g-sequencer-batch] up to block `C` must be included. Finally, the host program delivers the guest
```
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
The host program first executes the guest program to gather the necessary data, which is called [PreImageOracle] in [Optimism's Dispute Game]. For ZKFDG, however, additional data is required for PreImageOracle, which is the L1 block data from the origin of target L2 block to block `C`. Finally, the host program delivers the guest
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 10]
```diff
@@ -23,4 +37,91 @@
...
+      bytes32 outputRoot, 
+      bytes32 L1EndHash
+    );
+}
+```
+
+## Prover as an RPC Server
+
+`kroma-prover` is implemented as a jsonRPC server that generates a zkVM proof for the requested height block.
+It can be utilized as a local component on the challenger's machine or as a remote jsonRPC server (prover as a service).
```
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md]: I think we don't need this in specs.
```suggestion
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 11]
```diff
@@ -23,4 +37,91 @@
...
+
+#### Host Program
+
+The host program is a main part of the prover, responsible for delivering the guest program and its input to the zkVM.
+The host program first executes the guest program logic to gather the necessary data, which is equal to what is stored
+in the [PreImageOracle] during the [Optimism's Dispute Game], but the block headers from the L1 block containing
+the [sequencer-batch][g-sequencer-batch] up to block `C` must be included. Finally, the host program delivers the guest
+program and the preimages to the zkVM to obtain the zkVM proof.
+
+[PreImageOracle]: https://github.com/ethereum-optimism/specs/blob/main/specs/fault-proof/stage-one/fault-dispute-game.md#preimageoracle
```
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md]: I don't think it is sufficient as a spec of preimage oracle. Is there any detailed spec of preimage oracle for the guest program such as kona? It seems quite abstractive to me.

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 6 above)
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md]: Do we need to this section for ZKFDG spec? It's more like docs, not spec.

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 12]
```diff
@@ -23,4 +37,91 @@
 
 ## Proving Fault using ZKVM
 
+### Prover
+
+A prover plays a vital role in the challenge process, acting as an integral part of the challenger responsible
+for generating proof for the target block. This proof substantiates claims regarding the legitimacy of block data
+and state transitions.
+A [ZK fault proof][g-zk-fault-proof] demonstrates that a state transition from S to S’ is valid based on the
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md:39]: If you intended one line separation between these two paragraphs, an empty line should be added here.
```suggestion

A [ZK fault proof][g-zk-fault-proof] demonstrates that a state transition from S to S’ is valid based on the
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 13]
```diff
@@ -23,4 +37,91 @@
...
+transactions within the block. While this may seem similar to a validity proof, the key difference lies in its
+purpose. The ZK fault proof is used to demonstrate that a state transition from S to S’’ is incorrect by
+providing evidence of a valid state
+transition from S to S’.
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
providing evidence of a valid state transition from S to S’.
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 14]
```diff
@@ -23,4 +37,91 @@
...
+The zkVM (Zero-Knowledge Virtual Machine) is a virtual machine that executes guest program compiled with a specified
+compiler generating zero-knowledge proofs for their execution. The guest program can be written in standard programming
+languages, such as Rust and Go.
+
+#### Guest Program
+
+In the case of the ZK fault proof, the guest program is an extension of [L2 Chain Derivation][g-l2-chain-derivation]
+that includes a connectivity check among the blocks from L1 origin block and the specified block `C`. The hash of
+Block `C` is determined as the parent hash stored when the [Dispute Game][g-dispute-game] is created in Dispute Game
+Factory by calling `create()` of the Dispute Game Factory.
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
block `C` is determined as the parent hash stored when the ZK Fault Dispute Game is created by calling `create()` of
Dispute Game Factory.

```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 15]
```diff
@@ -23,4 +37,91 @@
...
+If the attacker manipulates any data within the extension of L2 Chain Derivation, it will affect the block hash `C`.
+This is because all data is ultimately linked to the block hash `C` through the hash chain. Therefore if any data
+manipulating attack be thwarted by checking the block hash `C` value. For example, if an attacker creates a proof
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
manipulating attack can be thwarted by checking the block hash `C` value. For example, if an attacker creates a proof
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 16]
```diff
@@ -23,4 +37,91 @@
...
+using a transaction that is not included in the [sequencer-batch][g-sequencer-batch], the block `C` value will change,
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
using a transaction that is not included in the [sequencer batch][g-sequencer-batch], the block `C` value will change,
```

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 9 above)
Comment (seolaoh) [reply specs/experimental/zk-fault-dipute-game/overview.md]: I think we don't use PreImageOracle, so better to remove that part.
```suggestion
The host program first executes the guest program to gather the necessary data, which is same as the data stored in [PreImageOracle] during [Optimism's Fault Dispute Game]. For ZKFDG, however, additional data is required, which is the L1 block data from the origin of target L2 block to block `C`. Finally, the host program delivers the guest
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 17]
```diff
@@ -23,4 +37,91 @@
...
+#### Host Program
+
+The host program is a main part of the prover, responsible for delivering the guest program and its input to the zkVM.
+The host program first executes the guest program logic to gather the necessary data, which is equal to what is stored
+in the [PreImageOracle] during the [Optimism's Dispute Game], but the block headers from the L1 block containing
+the [sequencer-batch][g-sequencer-batch] up to block `C` must be included. Finally, the host program delivers the guest
+program and the preimages to the zkVM to obtain the zkVM proof.
+
+[PreImageOracle]: https://github.com/ethereum-optimism/specs/blob/main/specs/fault-proof/stage-one/fault-dispute-game.md#preimageoracle
+[Optimism's Dispute Game]: https://github.com/ethereum-optimism/specs/blob/main/specs/fault-proof/stage-one/fault-dispute-game.md#fault-dispute-game
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]: It's better to fix the link to specific commit.
```suggestion
[PreImageOracle]: https://github.com/ethereum-optimism/specs/blob/46d411bfea922c520a1d43329dbf78a2f6966ae0/specs/fault-proof/stage-one/fault-dispute-game.md#preimageoracle
[Optimism's Dispute Game]: https://github.com/ethereum-optimism/specs/blob/46d411bfea922c520a1d43329dbf78a2f6966ae0/specs/fault-proof/stage-one/fault-dispute-game.md#fault-dispute-game
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 18]
```diff
@@ -23,4 +37,91 @@
...
+
+#### Public Values of Proof
+
+To mark which blocks have been executed, the proof publicly reveals the following data.
+
+``` plain
+1. Output Root at parent block of the target
+2. Output Root at the target block
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
1. Output root at the parent block
2. Output root at the target block
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 19]
```diff
@@ -23,4 +37,91 @@
...
+3. Hash of the block C
+```
+
+#### Zk Verify contract
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
#### `ZKVerifier` Interface
```

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 9 above)
Comment (0xHansLee) [reply specs/experimental/zk-fault-dipute-game/overview.md]: Actually, we use PreImageOracle off-chain (in prover), not on-chain. That's why @dongchangYoo add this part. The spec of PreImageOracle for guest program is the same with Optimism's. The only difference is the L1 Block info for validation of connectivity of L1 blocks.

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 20]
```diff
@@ -23,4 +37,91 @@
...
+
+The proof generated by a prover can be verified on a verifier contract that includes the following interface.
+The verification in a challenge process is implemented using the verify function provided by the verifier contract.
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
The verification in a challenge process will be implemented using the verify function provided by the verifier contract.
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 21]
```diff
@@ -23,4 +37,91 @@
...
+
+``` solidity
+interface ZKVerifier {
+    /// @notice The entrypoint for verifying the proof of zk fault proof.
+    /// @param _proofBytes The encoded proof.
+    /// @param _publicValues The encoded public values.
+    /// @return parentOutputRoot output root at parent block of the target.
+    /// @return outputRoot output root at the target block.
+    /// @return L1EndHash hash of the l1 end block.
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]:
```suggestion
    /// @param _publicValues The encoded public values.
    /// @param _proofBytes The encoded proof.
    /// @return parentOutputRoot The output root at the parent block.
    /// @return outputRoot The output root at the target block.
    /// @return L1EndHash The hash of the l1 end block.
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 22]
```diff
@@ -23,4 +37,91 @@
 
 ## Proving Fault using ZKVM
 
+### Prover
+
+A prover plays a vital role in the challenge process, acting as an integral part of the challenger responsible
+for generating proof for the target block. This proof substantiates claims regarding the legitimacy of block data
+and state transitions.
```
Comment (seolaoh) [inline specs/experimental/zk-fault-dipute-game/overview.md]: I'm not sure if Prover part is necessary for this section, Proving Fault using ZKVM. How about changing like below?
```suggestion
### ZK Fault Proof
```

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 14 above)
Comment (seolaoh) [reply specs/experimental/zk-fault-dipute-game/overview.md]: Also how about just adding DisputeGameFactory link of Optimism? Since we are planning to use the same contract.

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 9 above)
Comment (dongchangYoo) [reply specs/experimental/zk-fault-dipute-game/overview.md]: @0xHansLee Since `PreImageOracle` is a contract, not a data, i think that
```
the necessary data, which is equal to what is stored in the [PreImageOracle] during the [Optimism's Dispute Game]
```
is more accurate than
```
necessary data, which is called [PreImageOracle] in [Optimism's Dispute Game]
```

Comment (0xHansLee) [reply specs/experimental/zk-fault-dipute-game/overview.md]: Oh, I see. Let's go as you write.

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 6 above)
Comment (dongchangYoo) [reply specs/experimental/zk-fault-dipute-game/overview.md]: Do you mean that we should remove the 'Prover as an RPC Server' section?

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 11 above)
Comment (dongchangYoo) [reply specs/experimental/zk-fault-dipute-game/overview.md]: Does the Optimism spec provide a more detailed explanation of this data?

Comment (0xHansLee) [reply specs/experimental/zk-fault-dipute-game/overview.md]: I think more detailed spec for preimage data would be good than the link.

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 6 above)
Comment (0xHansLee) [reply specs/experimental/zk-fault-dipute-game/overview.md]: Yes, what do you think?

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 23]
```diff
@@ -21,6 +34,91 @@
...
+## Proving Fault using zkVM
+
+### ZK Fault Proof
+
+A [ZK fault proof][g-zk-fault-proof] demonstrates that a state transition from S to S’ is valid based on the
+transactions within the block. While this may seem similar to a validity proof, the key difference lies in its
+purpose. The ZK fault proof is used to demonstrate that a state transition from S to S’’ is incorrect by
+providing evidence of a valid state transition from S to S’.
+
+### zkVM Proof
```
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md]: This is more like related to zkVM, not zkVM proof imo
```suggestion
### zkVM
```

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 24]
```diff
@@ -21,6 +34,91 @@
...
+The verification in a challenge process will be implemented using the verify function provided by the verifier contract.
+
+``` solidity
+interface ZKVerifier {
+    /// @notice The entrypoint for verifying the proof of zk fault proof.
+    /// @param _publicValues The encoded public values.
+    /// @param _proofBytes The encoded proof.
+    /// @return parentOutputRoot The output root at the parent block of the target.
+    /// @return outputRoot The output root at the target block.
+    /// @return L1EndHash The hash of the l1 end block.
```
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md]: `L1EndHash` seems not to be clear. How about using just `l1BlockHash`? And describe what it is in description.
```suggestion
    /// @return l1BlockHash The block hash of L1 block stored in ZK fault dispute game contract.
```

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 11 above)
Comment (0xHansLee) [reply specs/experimental/zk-fault-dipute-game/overview.md]: If there is no, let's go as is.

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 24 above)
Comment (dongchangYoo) [reply specs/experimental/zk-fault-dipute-game/overview.md]: The term 'l1BlockHash' (the variable name) is too general. It would be better if it included a more specific meaning. So how about keeping it as it is?

Comment (0xHansLee) [reply specs/experimental/zk-fault-dipute-game/overview.md]: Ok, then how about `l1Head` as used in `FaultDisputeGame`? They used the term `l1Head` for the block hash.

https://github.com/ethereum-optimism/optimism/blob/0cf2a11611d329b9c6d53fd6473dbbc44f68e94c/packages/contracts-bedrock/src/dispute/interfaces/IDisputeGame.sol#L46-L46

File: specs/experimental/zk-fault-dipute-game/overview.md [hunk 25]
```diff
@@ -21,6 +33,79 @@
...
+The zkVM (Zero-Knowledge Virtual Machine) is a virtual machine that executes guest program compiled with a specified
+compiler generating zero-knowledge proofs for their execution. The guest program can be written in standard programming
+languages, such as Rust and Go.
+
+#### Guest Program
+
+For ZK Fault Dispute Game, the guest program is an extension of [L2 Chain Derivation][g-l2-chain-derivation] that
+includes a connectivity check among the blocks from L1 origin block to the specified block `C`. The hash of Block
+`C` is determined as the parent hash stored when the ZK Fault Dispute Game is created by calling `create()` of
+the [Dispute Game Factory].
```
Comment (0xHansLee) [inline specs/experimental/zk-fault-dipute-game/overview.md:55]: need link for this

Comment (dongchangYoo) [reply specs/experimental/zk-fault-dipute-game/overview.md:55]: The link is already located at the end of section!

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 24 above)
Comment (0xHansLee) [reply specs/experimental/zk-fault-dipute-game/overview.md]: @dongchangYoo please check this. Excepting this, all looks good to me.

**Review by 0xHansLee — APPROVED** [review]
Please check this [comment](https://github.com/kroma-network/kroma-specs/pull/19#discussion_r1740413911). Excepting this, LGTM

File: specs/experimental/zk-fault-dipute-game/overview.md (hunk 24 above)
Comment (dongchangYoo) [reply specs/experimental/zk-fault-dipute-game/overview.md]: It hash been changed.

**Review by 0xHansLee — APPROVED** [review]

**Review by seolaoh — APPROVED** [review]
LGTM
